contrib/babelfishpg_tsql/src/dbcmds.c

```
@@ -1231,3 +1259,279 @@ create_guest_schema_for_all_dbs(PG_FUNCTION_ARGS)
+
+		/* Run all subcommands */
+		foreach(parsetree_item, stmt_list)
+		{
+			Node		*stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+			PlannedStmt *wrapper;
+
+			/* need to make a wrapper PlannedStmt */
+			wrapper = makeNode(PlannedStmt);
+			wrapper->commandType = CMD_UTILITY;
+			wrapper->canSetTag = false;
+			wrapper->utilityStmt = stmt;
+			wrapper->stmt_location = 0;
+			wrapper->stmt_len = 0;
+
+			/* do this step */
+			ProcessUtility(wrapper,
+						"(CREATE FIXED DATABASE ROLES) ",
+						false,
+						PROCESS_UTILITY_SUBCOMMAND,
+						NULL,
+						NULL,
+						None_Receiver,
+						NULL);
+		}
+		tuple = heap_getnext(tblscan, ForwardScanDirection);
+	}
+
+	/* Cleanup. */
+	table_endscan(tblscan);
+	table_close(namespace_rel, RowExclusiveLock);
+	pfree(dbname);
+}
+
+static void
+create_db_roles_if_not_exists(const uint16 dbid,
+							const char *dbname)
+{
+	StringInfoData		query;
+	const char		*db_datareader;
+	const char 		*db_datawriter;
+	ListCell		*parsetree_item;
+	List			*stmt_list;
+	Node			*stmts;
+	int			i=0;
+	Oid			save_userid;
+	int			save_sec_context;
+
+	/*
+	 * During upgrade, the migration mode is reset to single-db so we cannot
+	 * call get_physical_user_name() directly. Detect whether the original
+	 * migration was single-db or multi-db.
+	 */
+	MigrationMode baseline_mode = is_user_database_singledb(dbname) ? SINGLE_DB : MULTI_DB;
+
+	db_datareader = get_db_datareader_name_by_mode(dbname, baseline_mode);
+	db_datawriter = get_db_datawriter_name_by_mode(dbname, baseline_mode);
+
+	/*
+	 * database roles prepends dbname based on single-db or multi-db. If for
+	 * some reason database roles exist, then that is a bigger problem.
+	 * The roles should be manually cleaned up before the upgrade.
+	 */
+	if (OidIsValid(get_role_oid(db_datareader, true)))
+	{
+		ereport(LOG,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("role \"%s\" already exists. Please drop the role and restart upgrade.", db_datareader)));
```

This should an ERROR right? Otherwise upgrade might silently go through without creating these essential users.

---

Oh right! It should be an ERROR.